### ISSUE TITLE
DYNARRAY and dangling pointers

### ISSUE THREAD

**Issue by dewyatt** [riboseinc/rnp#358]
## Description

Basically, I want to discourage the use of DYNARRAY in certain places and in certain ways because it has a good potential for leading to dangling pointers.

## Steps to Reproduce

As an example, it's pretty reasonable to expect something like this pseudocode would be fine:

```c
key_store = [...];
rnp_key_store_load_from_file([...], key-store);

key1 = rnp_key_store_get_key_by_id([...]);
key2 = rnp_key_store_get_key_by_id([...]);

// assume this loads additional keys into the same key store
rnp_key_store_load_from_file([...], key-store);

// any use of key1/key2 is potentially a segfault at this point
pgp_encrypt_file([...], key2);
```

The problem is that whenever EXPAND_ARRAY is called, if it triggers a realloc, all those pointers become invalid.

As a potential alternative, instead of something like:

```c
DYNARRAY(pgp_key_t, key);
```

You could do:

```c
DYNARRAY(pgp_key_t*, key);
```

Which is just a list of pointers, so you would need to individually allocate the structures, free them, etc.
You could also use the list implementation I introduced as well. Point is, we should probably try to avoid passing pointers around to API clients that may end up invalidated unexpectedly.

This hasn't caused any issues yet that I know of, but it will, and they could be a nightmare to troubleshoot.

**Comment by ronaldtse** [comment]
Agree with @dewyatt . We will need to check each and see.

Luckily there aren't too many of them:
```sh
include/rekey/rnp_key_store.h:    DYNARRAY(kbx_pgp_key_t, key);
include/rekey/rnp_key_store.h:    DYNARRAY(kbx_pgp_uid_t, uid);
include/rekey/rnp_key_store.h:    DYNARRAY(kbx_pgp_sig_t, sig);
include/rekey/rnp_key_store.h:    DYNARRAY(pgp_key_t, key);
include/rekey/rnp_key_store.h:    DYNARRAY(kbx_blob_t *, blob);
src/lib/packet.h:    DYNARRAY(uint8_t, symm_alg);
src/lib/packet.h:    DYNARRAY(uint8_t, hash_alg);
src/lib/packet.h:    DYNARRAY(uint8_t, compress_alg);
src/lib/packet.h:    DYNARRAY(uint8_t, key_server_pref);
src/lib/packet.h:    DYNARRAY(uint8_t *, uid);          /* array of user ids */
src/lib/packet.h:    DYNARRAY(pgp_rawpacket_t, packet); /* array of raw packets */
src/lib/packet.h:    DYNARRAY(pgp_subsig_t, subsig);    /* array of signature subkeys */
src/lib/packet.h:    DYNARRAY(pgp_revoke_t, revoke);    /* array of signature revocations */
src/librekey/key_store_g10.c:    DYNARRAY(s_exp_block_t, block);
src/librekey/key_store_g10.c:    DYNARRAY(s_exp_t, sub_s_exp);
```

**Comment by ni4** [comment]
Just hit into this problem with my new code, wasting some hours for debugging.

**Comment by dewyatt** [comment]
> Just hit into this problem with my new code, wasting some hours for debugging.

Yeah, honestly I've been thinking we may just want to completely remove DYNARRAY. The negatives of DYNARRAY, to me, are:

* It requires macros, which tend to be a bit more difficult to maintain (and currently uses fairly complex macros)
* It introduces the possibility of hard-to-debug issues like dangling pointers (this issue)
* There is no way to tell when it fails to expand. (This one isn't hard to fix, but currently if the realloc fails in EXPAND_ARRAY, there's no way for the caller to tell, except during the first alloc)

**Comment by ni4** [comment]
Agree on this, it add's more problems then solves. It could be replaced with functions like arr_init(int *size, int *vsize, int elsize); arr_resize(int *size, int *vsize, int elsize) and explicit declaration of those three fields in the struct.

**Comment by dewyatt** [comment]
> Agree on this, it add's more problems then solves. It could be replaced with functions like arr_init(int *size, int *vsize, int elsize); arr_resize(int *size, int *vsize, int elsize) and explicit declaration of those three fields in the struct.

This would be quite verbose IMO, but more importantly it would only solve the first issue if I understand correctly.

I just burned a few more hours on debugging a dynarray issue today. I realized that `rnp_key_store_remove_key` invalidates pointers (well, worse, it shifts the array of key data). Not that the function is wrong, but it's pretty much forced to behave that way because of dynarray.

I'm planning to evaluate `list` to replace (some?) uses of dynarray when time permits, but of course it does require some careful consideration.

**Comment by ronaldtse** [comment]
Thanks @ni4 !
